src/withApollo.tsx

```diff
@@ -22,66 +20,73 @@ export default function withApollo<TCache = any>(
 ) {
-  type ApolloProps = WithApolloProps<TCache> & AppProps;
+  type ApolloProps = WithApolloProps<TCache>;
 
   if (!options.getDataFromTree) {
-    options.getDataFromTree = 'always';
+    options.getDataFromTree = 'ssr';
   }
 
-  return (App: typeof NextApp) => {
-    return class WithApollo extends React.Component<ApolloProps> {
-      public static displayName = `WithApollo(${getDisplayName(App)})`;
+  return (Page: NextPage<any> | typeof App) => {
+    function WithApollo({ apollo, apolloState, ...props }: ApolloProps) {
+      const apolloClient =
+        apollo ||
+        initApollo<TCache>(client, { initialState: apolloState.data });
 
-      public static propTypes = {
-        apolloState: PropTypes.object,
-        apollo: PropTypes.object
-      };
+      return <Page {...props} apollo={apolloClient} />;
+    }
+    const getInitialProps = Page.getInitialProps;
+    const ssr = options.getDataFromTree === 'ssr';
+
+    WithApollo.displayName = `WithApollo(${getDisplayName(Page)})`;
 
-      public static getInitialProps = async (appCtx: ApolloContext) => {
-        const { AppTree, ctx } = appCtx;
+    if (getInitialProps || ssr) {
+      WithApollo.getInitialProps = async (pageCtx: ApolloContext) => {
+        const ctx = 'Component' in pageCtx ? pageCtx.ctx : pageCtx;
+        const { AppTree } = pageCtx;
         const headers = ctx.req ? ctx.req.headers : {};
         const apollo = initApollo<TCache>(client, { ctx, headers });
         const apolloState: WithApolloState<TCache> = {};
-        const getInitialProps = App.getInitialProps;
 
-        let appProps = { pageProps: {} };
+        let pageProps = {};
 
         if (getInitialProps) {
           ctx.apolloClient = apollo;
-          appProps = await getInitialProps(appCtx);
+          pageProps = await getInitialProps(pageCtx as any);
         }
 
-        if (ctx.res && (ctx.res.headersSent || ctx.res.finished)) {
-          return {};
-        }
+        if (typeof window === 'undefined') {
+          if (ctx.res && (ctx.res.headersSent || ctx.res.finished)) {
+            return pageProps;
+          }
 
-        if (
-          options.getDataFromTree === 'always' ||
-          (options.getDataFromTree === 'ssr' && typeof window === 'undefined')
-        ) {
-          try {
-            await getDataFromTree(
-              <AppTree
-                {...appProps}
-                apolloState={apolloState}
-                apollo={apollo}
-              />
-            );
-          } catch (error) {
-            // Prevent Apollo Client GraphQL errors from crashing SSR.
-            if (process.env.NODE_ENV !== 'production') {
-              // tslint:disable-next-line no-console This is a necessary debugging log
-              console.error('GraphQL error occurred [getDataFromTree]', error);
+          if (ssr) {
+            try {
+              const { getDataFromTree } = await import('@apollo/react-ssr');
+
+              await getDataFromTree(
```

Comment:
It is recommended to use `getMarkupFromTree` ?
refs  https://github.com/apollographql/react-apollo/issues/3251#issuecomment-513223453

Reply:
Where are the docs for that?